nomad/worker.go

```diff
@@ -215,9 +215,20 @@ func (w *Worker) waitForIndex(index uint64, timeout time.Duration) error {
 	start := time.Now()
 	defer metrics.MeasureSince([]string{"nomad", "worker", "wait_for_index"}, start)
 CHECK:
+	// Snapshot the current state
+	snap, err := w.srv.fsm.State().Snapshot()
```

Comment:
One other thought I had when looking at this was could you use your state store's blocking query watch support for this? That way you get an edge to wake this up with when the data arrives vs. letting the wait expire.